test/dynamax.c

```diff
@@ -0,0 +1,1386 @@
+    } FINALLY {
+        EXPECT_MUL_EQ(results[0].damage, UQ_4_12(0.25), results[1].damage);
+    }
+}
+
+SINGLE_BATTLE_TEST("(DYNAMAX) Max Moves don't bypass Max Guard")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_PROTECT, dynamax: TRUE); }
+    } SCENE {
+        NONE_OF { HP_BAR(opponent); }
+    }
+}
+
+DOUBLE_BATTLE_TEST("(DYNAMAX) Feint bypasses Max Guard but doesn't break it")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        PLAYER(SPECIES_WYNAUT);
+        OPPONENT(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WYNAUT);
+    } WHEN {
+        TURN { MOVE(playerLeft, MOVE_PROTECT, dynamax: TRUE);
+               MOVE(opponentLeft, MOVE_FEINT, target: playerLeft);
+               MOVE(opponentRight, MOVE_TACKLE, target: playerLeft);
+        }
+    } SCENE {
+        MESSAGE("Wobbuffet used Max Guard!");
+        MESSAGE("Foe Wobbuffet used Feint!");
+        HP_BAR(playerLeft);
+        MESSAGE("Foe Wynaut used Tackle!");
+        NONE_OF { HP_BAR(playerLeft); }
+    }
+}
+
+DOUBLE_BATTLE_TEST("(DYNAMAX) Dynamaxed Pokemon are immune to Instruct")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        PLAYER(SPECIES_WYNAUT);
+        OPPONENT(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WYNAUT);
+    } WHEN {
+        TURN { MOVE(playerLeft, MOVE_TACKLE, dynamax: TRUE, target: opponentLeft);
+               MOVE(playerRight, MOVE_INSTRUCT, target: playerLeft);
+        }
+    } SCENE {
+        MESSAGE("Wobbuffet used Max Strike!");
+        MESSAGE("Wynaut used Instruct!");
+        MESSAGE("But it failed!");
+    }
+}
+
+// TODO: Gigantamax factor
+SINGLE_BATTLE_TEST("(DYNAMAX) Pokemon with Gigantamax forms change upon Dynamaxing")
+{
+    GIVEN {
+        PLAYER(SPECIES_VENUSAUR);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); }
+    } FINALLY {
+        EXPECT_EQ(gBattleMons[B_POSITION_PLAYER_LEFT].species, SPECIES_VENUSAUR_GMAX);
+    }
+}
+
+SINGLE_BATTLE_TEST("(DYNAMAX) Pokemon with Gigantamax forms revert upon switching")
+{
+    GIVEN {
+        PLAYER(SPECIES_VENUSAUR);
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); }
+        TURN { SWITCH(player, 1); }
+        TURN { SWITCH(player, 0); }
+    } FINALLY {
+        EXPECT_EQ(gBattleMons[B_POSITION_PLAYER_LEFT].species, SPECIES_VENUSAUR);
+    }
+}
+
+SINGLE_BATTLE_TEST("(DYNAMAX) Pokemon with Gigantamax forms revert upon fainting")
+{
+    GIVEN {
+        PLAYER(SPECIES_VENUSAUR) { HP(1); };
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE, dynamax: TRUE); MOVE(opponent, MOVE_TACKLE); }
+    } FINALLY {
+        EXPECT_EQ(gBattleMons[B_POSITION_PLAYER_LEFT].species, SPECIES_VENUSAUR);
```

Comment:
```suggestion
    } THEN {
        EXPECT_EQ(player->species, SPECIES_VENUSAUR);
```
can we even check the player form at this point?